pgcli/packages/function_metadata.py

```diff
@@ -0,0 +1,146 @@
+                    self.is_set_returning))
+
+    def __repr__(self):
+        return (('%s(schema_name=%r, func_name=%r, arg_list=%r, return_type=%r,'
+                 ' is_aggregate=%r, is_window=%r, is_set_returning=%r)')
+                % (self.__class__.__name__, self.schema_name, self.func_name,
+                   self.arg_list, self.return_type, self.is_aggregate,
+                   self.is_window, self.is_set_returning))
+
+    def fieldnames(self):
+        """Returns a list of output field names"""
+
+        if self.return_type.lower() == 'void':
+            return []
+
+        match = table_def_regex.match(self.return_type)
+        if match:
+            # Function returns a table -- get the column names
+            return list(field_names(match.group(1), mode_filter=None))
+
+        # Function may have named output arguments -- find them and return
+        # their names
+        return list(field_names(self.arg_list, mode_filter=('OUT', 'INOUT')))
+
+
+class TypedFieldMetadata(object):
+    """Describes typed field from a function signature or table definition
+
+        Attributes are:
+            name        The name of the argument/column
+            mode        'IN', 'OUT', 'INOUT', 'VARIADIC'
+            type        A list of tokens denoting the type
+            default     A list of tokens denoting the default value
+            unknown     A list of tokens not assigned to type or default
+    """
+
+    __slots__ = ['name', 'mode', 'type', 'default', 'unknown']
+
+    def __init__(self):
+        self.name = None
+        self.mode = 'IN'
+        self.type = []
+        self.default = []
+        self.unknown = []
+
+
+def parse_typed_field_list(tokens):
+    """Parses a argument/column list, yielding TypedFieldMetadata objects
+
+        Field/column lists are used in function signatures and table
+        definitions. This function parses a flattened list of sqlparse tokens
+        and yields one metadata argument per argument / column.
+    """
+
+    # postgres function argument list syntax:
+    #   " ( [ [ argmode ] [ argname ] argtype
+    #               [ { DEFAULT | = } default_expr ] [, ...] ] )"
+
+    mode_names = set(('IN', 'OUT', 'INOUT', 'VARIADIC'))
+    parse_state = 'type'
+    parens = 0
+    field = TypedFieldMetadata()
+
+    for tok in tokens:
+        if tok.ttype in Whitespace or tok.ttype in Comment:
+            continue
+        elif tok.ttype in Punctuation:
+            if parens == 0 and tok.value == ',':
+                # End of the current field specification
+                if field.type:
+                    yield field
+                # Initialize metadata holder for the next field
+                field, parse_state = TypedFieldMetadata(), 'type'
+            elif parens == 0 and tok.value == '=':
+                parse_state = 'default'
+            else:
+                getattr(field, parse_state).append(tok)
```

Comment:
Is there a reason for using `getattr(field, parse_state).append(tok)` instead of `field[parse_state].append(tok)`. 

The second one reads better (personal opinion, so feel free to disagree).